src/ExtraRegisters.cc

```diff
@@ -596,10 +596,12 @@ void ExtraRegisters::reset() {
       * Avoid this issue by setting the bit if the feature is supported by the
       * CPU.
       */
-      const XSaveLayout& native_layout = xsave_native_layout();
       uint64_t pkru_bit = uint64_t(1) << xsave_feature_pkru;
-      if (native_layout.supported_feature_bits & pkru_bit) {
-        xinuse |= pkru_bit;
+      if (xcr0() & pkru_bit) {
+        const XSaveLayout& native_layout = xsave_native_layout();
+        if (native_layout.supported_feature_bits & pkru_bit) {
```

Comment:
I don't think we need the native_layout check as well, do we? If the bit is enabled in XCR0 it had better be enabled in CPUID.

Reply:
But the native_layout is not directly related to cpuid, is it?

Reply:
`xsave_native_layout().supported_feature_bits` is basically just the result of `cpuid(CPUID_GETXSAVE, 0)`. We could probably just call the latter instead, if we wanted to, but I think we can just get rid of it.

Reply:
The SDM says, for CPUID EAX=0x0d (CPUID_GETXSAVE), ECX=0:
> EAX Bits 31 - 00: Reports the supported bits of the lower 32 bits of XCR0. XCR0[n] can be set to 1 only if EAX[n] is 1.
So if we find that the PKRU bit is set in XCR0 but it is not set in CPUID_GETXSAVE/0 (i.e. `xsave_native_layout().supported_feature_bits`), then something has gone terribly wrong.

Reply:
Fair enough.